apps/coordinator/ts/scheduler/scheduler.service.ts

```diff
@@ -0,0 +1,292 @@
+    const key = getPollKeyFromObject(scheduledPoll);
+    const { maciAddress, pollId, chain } = scheduledPoll;
+
+    const storedPoll = await this.redisService.get(key);
+
+    if (!storedPoll) {
+      return { isScheduled: false };
+    }
+
+    try {
+      this.schedulerRegistry.getTimeout(key);
+    } catch (error) {
+      // delete redis data so user has to reschedule again
+      await this.redisService.delete(key);
+      return { isScheduled: false };
+    }
+
+    const { isPollTallied } = await this.getPollFinalizationData({
+      maciAddress,
+      pollId,
+      chain,
+    });
+
+    if (isPollTallied) {
+      this.deleteScheduledPoll({ maciAddress, pollId, chain });
+      this.logger.error(`Error: ${ErrorCodes.POLL_ALREADY_TALLIED}, poll has already been tallied`);
+      throw new Error(ErrorCodes.POLL_ALREADY_TALLIED.toString());
+    }
+
+    return { isScheduled: true };
+  }
+
+  /**
+   * Return the poll end date and whether it is tallied
+   * @param args - Arguments to retrieve poll data
+   */
+  async getPollFinalizationData({
+    pollId,
+    maciAddress,
+    chain,
+    sessionKeyAddress,
+    approval,
+  }: IIdentityPollWithSignerArgs): Promise<IGetPollFinalizationData> {
+    const signer = await this.sessionKeysService.getCoordinatorSigner(chain, sessionKeyAddress, approval);
+
+    const [{ endDate, isMerged }, isPollTallied] = await Promise.all([
+      getPoll({
+        maciAddress,
+        pollId: BigInt(pollId),
+        signer,
+      }),
+      isTallied({
+        maciAddress,
+        pollId: String(pollId),
+        signer,
+      }),
+    ]);
+
+    return {
+      endDate: Number(endDate),
+      isPollTallied: isMerged ? isPollTallied : false,
+    };
+  }
+
+  /**
+   * Store poll information and schedule its finalization.
+   * @param args - Arguments for storing and scheduling poll finalization
+   * @return delay in milliseconds until the poll is scheduled for finalization
+   */
+  private async setupPollFinalization({
+    maciAddress,
+    pollId,
+    mode,
+    deploymentBlockNumber,
+    chain,
+    endDate,
+  }: IScheduledPoll): Promise<ISetupPollFinalizationResponse> {
+    const key = getPollKeyFromObject({ maciAddress, pollId, chain });
+
+    await this.redisService.set(
+      key,
+      JSON.stringify({
+        maciAddress,
+        pollId,
+        mode,
+        deploymentBlockNumber,
+        chain,
+        endDate,
+      } as IScheduledPoll),
+    );
+
+    const shouldBeExecutedIn = endDate * 1000 - Date.now();
+    const toBeExecutedIn = shouldBeExecutedIn > 0 ? shouldBeExecutedIn : 0;
+    const delay = toBeExecutedIn + BUFFER_TIMEOUT;
+
+    const timeout = setTimeout(() => {
+      // TODO: execute this.finalizePoll (implementing this in the next PR)
+      this.redisService.delete(key);
+    }, delay);
+
+    const doesExist = this.schedulerRegistry.doesExist("timeout", key);
+    if (doesExist) {
```

Comment:
And it should have different name because variable should not be a verb

```suggestion
    const doesExist = this.schedulerRegistry.doesExist("timeout", key);

    if (doesExist) {
```